### PR TITLE
Add type in the response for RUN

### PR DESCRIPTION
**Pull request by pontusmelke**
Instead of only return `type: [r,w,rw]` in response to
`PULL_ALL` also include it in the response of `RUN`.

### CODE REVIEW

**Comment by boggle** [comment]
I like including the type; just surprised we know it that early (i.e. before sending data). This wouldn't require a change in the client API, i.e. the java driver would still just call `result.summarize().statementType()` to get it, right?

**Comment by pontusmelke** [comment]
@boggle This is not before sending any data right? We figure out the query type when parsing the query. and when we do `RUN` we have already parsed and created the a `org.neo4j.graphdb.Result` which contains both the `fields` and the `QueryExecutionType` - we just haven't consumed the actual data yet.

The type is still included in the `PULL_ALL` response so the API shouldn't have to change. The driver could figure out the type without actually consuming the result though. So if we like we could add `statementType` directly to `Result` in the driver, e.g. `result.statementType()`.

**Comment by zhenlineo** [comment]
I think boggle is talking about before streaming any result back. All in all +1 to this idea

File: community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/CypherAdapterStream.java [hunk 1]
```diff
@@ -72,7 +80,7 @@ public boolean visit( Result.ResultRow row ) throws Exception
         } );
 
         QueryExecutionType qt = delegate.getQueryExecutionType();
-        visitor.addMetadata( "type", queryTypeCode( qt.queryType() ) );
+        visitor.addMetadata( "type", queryType );
```
Comment (zhenlineo) [inline community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/CypherAdapterStream.java:83]: Are we still sending `type` in the `SUCCESS` for `PULL_ALL`? Shall we just remove it? 
Then we also need to modify in drivers to save `type` info when they got the `SUCCESS` for `RUN`

File: community/bolt/src/main/java/org/neo4j/bolt/v1/messaging/msgprocess/RunCallback.java [hunk 2]
```diff
@@ -38,6 +38,10 @@ public RunCallback( Log log )
     public void result( StatementMetadata result, Void none ) throws Exception
     {
         successMetadata.put( "fields", result.fieldNames() );
+        if ( !result.queryType().isEmpty() )
```
Comment (zhenlineo) [inline community/bolt/src/main/java/org/neo4j/bolt/v1/messaging/msgprocess/RunCallback.java:41]: Just curious, can we get a cypher query that we cannot deduce the query type after we execute/parse it?

File: community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/CypherAdapterStream.java (hunk 1 above)
Comment (pontusmelke) [inline community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/CypherAdapterStream.java:83]: yes, I didn't touch that. I can remove it if that is what we think that is better? 

Just to clarify I was also talking about before streaming the result back. We know the `type` after parsing the query, before streaming the result so it should not be a problem of providing the `type` as a response to `RUN`

File: community/bolt/src/main/java/org/neo4j/bolt/v1/messaging/msgprocess/RunCallback.java (hunk 2 above)
Comment (pontusmelke) [inline community/bolt/src/main/java/org/neo4j/bolt/v1/messaging/msgprocess/RunCallback.java:41]: For example when  you do `ROLLBACK` we return a `RecordStream.EMPTY`

File: community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/CypherAdapterStream.java (hunk 1 above)
Comment (zhenlineo) [inline community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/CypherAdapterStream.java:83]: I think it is fair to remove `type` from `PULL_ALL` as there is no need to stream to client the same info twice

**Comment by jakewins** [comment]
I'm not entirely sure of the use case for this one - just because we have the information available earlier does not strictly mean we should provide it earlier. It does kind of go nicely with the result column metadata in the stream header though.. Still, this would require equivalent changes in all the drivers, so I'd suggest we wait with this modification until a later protocol version where we enshrine in a stricter way what goes in the stream headers and footers.

**Pull request closed without merging**